README.md
# RadialSlider

The `RadialSlider` is an angular slider widget, which acts similarly to
the kivy's native linear slider.
The RadialSlider has two components: a **track** and a **thumb**. The thumb → ●, can be dragged along the circular path of the track → ◯, with an angle
ranging from 0 to 360 degrees.

You can define the position of the **thumb** via the `angle`, the `value`, or the `normalized_value`. 
Note that these values are also automatically updated as the thumb is dragged.

![Demo](doc/images/demo.gif?raw=true "Demo")

[![Github Build Status](https://github.com/kivy-garden/flower/workflows/Garden%20flower/badge.svg)](https://github.com/kivy-garden/flower/actions)

## Install
	pip install kivy_garden.radialslider

## Tips and Considerations:

### Widget size setting

The Radial Slider will adjust to the smallest size, between height and width.
Setting different sizes for the width and height is not recommended, as although the visible path and thumb will adjust to the smallest size, the widget as a whole will still conform to the size you set.

While setting the size incorrectly does not compromise the functionality of the widget itself, you might be overwriting another widget's area resulting in unexpected behavior of the other widget.

The example bellow shows the difference between the unespected behavior with wrong fitting and the correct fitting. The red area shows the real widget size.

Wrong size definition             |Wrong size definition             |  Correct size definition
:-------------------------|:-------------------------|:-------------------------
![Correct size setting](doc/images/wrong_width.png?raw=true "Wrong")  | ![Correct size setting](doc/images/wrong_height.png?raw=true "Wrong")  |  ![Correct size setting](doc/images/correct.png?raw=true "Correct")
`size_hint: (None, None)` <br> `size: 250, 200` ❌	|`size_hint: (None, None)` <br> `size: 200, 250` ❌	| `size_hint: (None, None)` <br> `size: 200, 200` ✔️


If you want to set the size based on `size_hint`, the proper way would be:

<br>

For adjust the height to the width:
```yaml
size_hint: (your_size_hint_x, None)
height: self.width
```

<br>

Or for adjust the width to the height:
```yaml
size_hint: (None, your_size_hint_y)
width: self.height
```

<br>

Or just using the `size` property:
```yaml
size_hint: (None, None)
size: your_width, your_height # one value must be the same as the other
```

<br>

🟡 If the parent of the RadialSlider widget has equal width and height dimensions, eg: `size: (200, 200)`. You can use size `size_hint_x` and `size_hint_y` freely, **as long as you set the same value for both**, like: `size_hint:(0.5, 0.5 )`, `size_hint:(0.2, 0.2)`, etc.

## Example:

The RadialSlider widget provides the essential components: *Thumb* → ●, and *Track* → ◯.

But with a few lines of code, you can change its appearance as you wish. Take a look at the example below, how easy it is.

![Example](doc/images/example.gif?raw=true "Correct")

```python
RadialSlider:
    size_hint: (None, None)
    size: 200, 200
    track_color: "#DEDEDE" # gray track
    track_thickness: 15
    canvas.before:
        # white background
        Color:
            rgba: rgba("#FFFFFF")
        Ellipse:
            size: self.size
            pos: self.pos
    canvas:
    	# blue line
        Color:
            rgba: rgba("#3086BD")
        Line:
            width: 15
            circle: self.center_x, self.center_y, self.width/2 - 15, 0, self.angle
            cap_precision: 500
    Label:
        pos: self.parent.center_x - self.width/2, self.parent.center_y - self.height/2
        text: "{}%".format(int(self.parent.value))
        color: "#808080"
        font_size: dp(35)
```
---
## API

	angle

> Current angle used for the Radial Slider. You can use it for setting
> the angle of the thumb in the track in range 0 - 360.
> 
> `angle` is defaults to 0.  

<br/>

	max_value

> Maximum value allowed for the Radial Slider.
> 
> `max_value` is defaults to 1.

<br/>

	min_value

> Minimum value allowed for the Radial Slider.
> 
> `min_value` is defaults to 0.

<br/>

	thumb_color

> Color of the thumb.
> 
> `thumb_color` is defaults to '#ffffff.'''

<br/>

	thumb_diameter

> Diameter of the thumb.
> 
> `thumb_diameter` is defaults to 25.

<br/>

	track_color

> Color of the track.
> 
> `track_color` is defaults to "#ffffff".

<br/>

	track_thickness

> Thickness of the track.
> 
> `track_thickness`is defaults to 2.'''

<br/>

	value

> Normalized value inside the range (min_value - max_value). You can use
> it for setting the value betwwen the minimum and maximum value.
> 
> `value` is defaults to 0.

<br/>

	value_normalized

> Normalized value inside the range (min_value - max_value) to 0-1
> range. You can also use it for setting the real value without knowing
> the minimum and maximum value.
> 
> `value_normalized` is defaults to 0
---
CI
--

Every push or pull request run the [GitHub Action](https://github.com/kivy-garden/flower/actions) CI.
It tests the code on various OS and also generates wheels that can be released on PyPI upon a
tag. Docs are also generated and uploaded to the repo as well as artifacts of the CI.


TODO
-------

* add the option to control the start and end angle of the RadialSlider, keeping the current behavior.
* create a new class/widget called RadialRangeSlider, which makes range selection possible.

Contributing
--------------

Check out our [contribution guide](CONTRIBUTING.md) and feel free to improve the RadialSlider flower.

🔴 If you have fulfilled any of the TODO items, feel free to contribute to the community by opening a pull request.

🟠 If you have a bug or an idea, create a report to help us improve or suggest an idea for this project by opening an issue

License
---------

This software is released under the terms of the MIT License.
Please see the [LICENSE.txt](LICENSE.txt) file.



CHANGELOG.md
# Changelog

## [v0.1.1] - 2021-08-12
 - Initial release.


kivy_garden/radialslider/__init__.py
"""
Radial Slider
============

The :class:`RadialSlider` is an angular slider widget, which acts similarly to
the linear slider. Thumb movement is limited to a circular path, with an angle
ranging from 0 to 360 degrees

The radial slider has two components: a track and a thumb. The thumb can be
dragged along the circular path of the track.

You can define the position of the thumb via the angle, the value, or the
normalized value. These values are also automatically updated as the thumb is
dragged.

"""

__all__ = ('RadialSlider', )

from ._version import __version__

from math import sin, cos, atan2, radians
from kivy.uix.widget import Widget
from kivy.graphics import Color, SmoothLine, Ellipse
from kivy.properties import NumericProperty, ColorProperty, AliasProperty


class RadialSlider(Widget):
    '''A :class:`~kivy.uix.widget.Widget` that provides a radial slider.
    '''

    angle = NumericProperty(0)
    '''Current angle used for the Radial Slider.

    You can use it for setting the angle of the thumb in the track in range
    0 - 360.

    :attr:`angle` is a :class:`~kivy.properties.NumericProperty` and defaults
    to 0.'''

    track_color = ColorProperty('#ffffff')
    '''Color of the track.

    :attr:`track_color` is a :class:`~kivy.properties.ColorProperty` and
    defaults to '#ffffff.'''

    thumb_color = ColorProperty('#ffffff')
    '''Color of the thumb.

    :attr:`thumb_color` is a :class:`~kivy.properties.ColorProperty` and
    defaults to '#ffffff.'''

    thumb_diameter = NumericProperty(10)
    '''Diameter of the thumb.

    :attr:`thumb_diameter` is a :class:`~kivy.properties.NumericProperty` and
    defaults to 10.'''

    track_thickness = NumericProperty(2)
    '''Thickness of the track.

    :attr:`track_thickness` is a :class:`~kivy.properties.NumericProperty` and
    defaults to 2.'''

    max_value = NumericProperty(100)
    '''Maximum value allowed for the Radial Slider.

    :attr:`max_value` is a :class:`~kivy.properties.NumericProperty` and
    defaults to 1.'''

    min_value = NumericProperty(0)
    '''Minimum value allowed for the Radial Slider.

    :attr:`min_value` is a :class:`~kivy.properties.NumericProperty` and
    defaults to 0.'''

    def get_value(self):
        value_range = abs(self.min_value - self.max_value)
        return self.min_value + max(0, min(self.angle, 360))/360 * value_range

    def set_value(self, value):
        self.angle = max(0, min((value + abs(self.min_value))/abs(
            self.max_value - self.min_value) * 360, 360))

    value = AliasProperty(get_value, set_value,
                          bind=('angle', 'min_value', 'max_value'),
                          cache=True)
    '''Normalized value inside the :attr:`range` (min_value/max_value).

    You can use it for setting the value betwwen the minimum and maximum value.

    :attr:`value_normalized` is an :class:`~kivy.properties.AliasProperty`.
    '''

    def get_norm_value(self):
        return self.angle/360

    def set_norm_value(self, value):
        self.angle = value * 360

    value_normalized = AliasProperty(get_norm_value, set_norm_value,
                                     bind=('value', 'min_value', 'max_value',
                                           'angle'),
                                     cache=True)
    '''Normalized value inside the :attr:`range` (min_value/max_value) to
    0-1 range.

    You can also use it for setting the real value without knowing the minimum
    and maximum value.

    :attr:`value_normalized` is an :class:`~kivy.properties.AliasProperty`.
    '''

    def __init__(self, **kwargs):
        super(RadialSlider, self).__init__(**kwargs)
        self._start_angle = 0
        self._end_angle = 360
        self._touch_control = None

        with self.canvas:
            self._track_color = Color(1, 1, 1, 1)
            self._track = SmoothLine(width=2, overdraw_width=2.5,
                                     circle=(self.center_x, self.center_y,
                                             0.5*self.width, 0, 360, 800))
        with self.canvas.after:
            self._thumb_color = Color(1, 1, 1, 1)
            self._thumb = Ellipse(size=self.size, pos=self.pos)
            self._thumb_border = SmoothLine(width=1,
                                            overdraw_width=(2.5),
                                            circle=(self._thumb.pos[0],
                                                    self._thumb.pos[1],
                                                    (self.thumb_diameter/2),
                                                    0, 360, 800))

        self.bind(size=self.update_canvas, pos=self.update_canvas,
                  angle=self.update_canvas)

    def update_canvas(self, instance, value):
        self._track_color.rgba = self.track_color
        self._track.width = self.track_thickness

        if (self.track_thickness < self.thumb_diameter):
            track_adjustment = self.thumb_diameter
        else:
            track_adjustment = self.track_thickness

        min_dimension = min(self.width/2, self.height/2)

        self._track.circle = (self.center_x, self.center_y, min_dimension -
                              track_adjustment, self._start_angle,
                              self._end_angle, 800)

        self._thumb_color.rgba = self.thumb_color
        self._thumb.size = self.thumb_diameter * 2, self.thumb_diameter * 2

        angle = 90 - max(self._start_angle, min(self.angle, self._end_angle))
        spacing = self.track_thickness - self.thumb_diameter
        spacing = 0 if spacing < 0 else spacing
        self._thumb.pos = (self.center_x + (min_dimension -
                           self.thumb_diameter) * cos(radians(angle)) -
                           (self.thumb_diameter) - spacing * (cos(radians(
                            angle))), self.center_y + (min_dimension -
                           (self.thumb_diameter)) * sin(radians(angle)) -
                           (self.thumb_diameter) - spacing * (sin(radians(
                            angle))))

        self._thumb_border.circle = (self._thumb.pos[0] + (
            self.thumb_diameter), self._thumb.pos[1] + (self.thumb_diameter),
            (self.thumb_diameter), 0, 360, 800)

    def on_touch_down(self, touch):
        if self.disabled or self._touch_control is not None:
            return
        if not self.collide_point(*touch.pos):
            return

        # check if the touch is in the dragable thumb
        thumb_limits = (touch.x >= self._thumb.pos[0],
                        touch.x <= self._thumb.pos[0] + self._thumb.size[0],
                        touch.y >= self._thumb.pos[1],
                        touch.y <= self._thumb.pos[1] + self._thumb.size[1])
        if not all(thumb_limits):
            return

        touch.grab(self)
        self._touch_control = touch
        return True

    def on_touch_move(self, touch):
        if touch.grab_current is not self:
            return
        angle = self.get_angle(touch.x, touch.y) * 180 / (3.14)

        if self._start_angle != 0 or self._end_angle != 360:
            self.angle = self.get_angle_norm(angle)
        else:
            # avoid overscrolling to 0
            if (touch.x + 1 > self.center_x) and self.angle > 270:
                self.angle = 360
            # avoid overscrolling to 360
            elif (touch.x - 1 <= self.center_x) and self.angle < 90:
                self.angle = 0
            else:
                self.angle = self.get_angle_norm(angle)

        return True

    def on_touch_up(self, touch):
        if touch.grab_current is not self:
            return
        touch.ungrab(self)
        self._touch_control = None
        return True

    def get_angle(self, x, y):
        cx, cy = self.center
        delta_x = x - cx
        delta_y = y - cy
        return atan2(delta_y, delta_x)

    # current angle in range 1 - 360
    def get_angle_norm(self, angle):
        if round(angle) >= 90 and round(angle) <= 180:
            return round(360 + (90 - angle))
        else:
            return round(90 - angle)
